docs/design/proposals/encryption-with-vault-tokens.md

```diff
@@ -0,0 +1,109 @@
+documented](./encrypted-pvc.md):
+
+- tenants can create PVCs that are encrypted with a unique key
+- the key to encrypt/decrypt the PVC is stored in Hashicorp Vault
+- there is a single token (Kubernetes ServiceAccount) to access Vault
+- enabling is done in the StorageClass, pointing to a KMS configuration section
+  in a JSON configuration file (Kubernetes ConfigMap)
+
+## High-Level Requirements
+
+The new feature should support multi-tenancy, so that each Tenant can use their
+own Vault Token for PVC encryption:
+
+- each tenant can manage their Vault token (and PVC keys)
+- to get the key from Vault, each tenant should provide their personal token
+- Ceph-CSI should use the token from the tenant to store the unique key for the
+  PVC
+- a Tenant is the owner of a Kubernetes Namespace
+
+## Restrictions to consider
+
+- tenants can only configure their Kubernetes namespace, their token to access
+  Vault should be located in their namespace (as a Kubernetes Secret)
+- Ceph-CSI can not directly access Kubernetes namespaces, as CSI is an
+  abstraction layer
+- Ceph-CSI needs to talk to a service (sidecar) to access the token from a
+  tenant
+- the KMS configuration is a ConfigMap with name
+  `ceph-csi-encryption-kms-config` in the namespace where the Ceph-CSI pods are
+  running
+- the KMS configuration is available for the Ceph-CSI pods at
+  `/etc/ceph-csi-encryption-kms-config/config.json`
+- [example of the
+  configuration](https://github.com/ceph/ceph-csi/blob/master/examples/kms/vault/kms-config.yaml)
+
+## Dependencies
+
+- the name of the Kubernetes Secret needs to be known (configured in the
+  StorageClass, or somewhere else)
+- each tenant (Kubernetes namespace) should use the same name for the Secret
+  that contains the Vault token
+- the Ceph-CSI sidecar needs to store the namespace (and name of the Secret?)
+  together with the metadata of the PVC
+
+## Implementation Outline
+
+- the Ceph-CSI node-plugin needs to request the token in the NodeStage CSI
+  operation and create/get the key for the PVC
+- the Ceph-CSI provisioner needs to request the token to delete the key for the
+  PVC in the VolumeDelete CSI operation
+- when creating the PVC the Ceph-CSI provisioner needs to store the namespace
+  (and Kubernetes Secret?) of the PVC in its metadata
+  - stores the `csi.volume.owner` (name of Tenant) in the journal and sets it
+    as `rbdVolume.Owner`
+
+## Configuration Details
+
+- the current KMS configuration file needs extensions for Vault Token support
+- introduce a new KMS-type: VaultTokenKMS (VaultKMS uses a Kubernetes
+  ServiceAccount)
+- configuration of the VaultTokenKMS can be very similar to VaultKMS for common
+  settings
+- tokens for each tenant can be added to the configuration (insecure & needs
+  admin involvement)
+- the configuration can describe defaults for each tenant
+  - Vault Service connection details (address, TLS options, ...)
+  - default token (in case of a single token for every tenant)
```

Comment:
the default token is also stored in the secret, the default secret name needs to be specified here?

Reply:
There does not need to be a default (global/shared) token. Everything that can be configured in a pre-Tenant secret, can also be configured in the KMS configuration file.